tooling/debugger/src/repl.rs

```diff
@@ -121,9 +137,76 @@ impl<'a, B: BlackBoxFunctionSolver> ReplDebugger<'a, B> {
         }
     }
 
+    fn display_opcodes(&self) {
+        let opcodes = self.context.get_opcodes();
+        let current_opcode_location = self.context.get_current_opcode_location();
+        let current_ip = match current_opcode_location {
+            Some(OpcodeLocation::Acir(ip)) => Some(ip),
+            Some(OpcodeLocation::Brillig { acir_index, .. }) => Some(acir_index),
+            None => None,
+        };
+        let current_brillig_pc = match current_opcode_location {
+            Some(OpcodeLocation::Brillig { brillig_index, .. }) => brillig_index,
+            _ => 0,
+        };
+        for (ip, opcode) in opcodes.iter().enumerate() {
+            let marker = if current_ip == Some(ip) {
+                "->"
+            } else if self.context.is_breakpoint_set(&OpcodeLocation::Acir(ip)) {
+                " *"
+            } else {
+                ""
+            };
+            if let Opcode::Brillig(brillig) = opcode {
+                println!("{:>3} {:2} BRILLIG inputs={:?}", ip, marker, brillig.inputs);
+                println!("       |       outputs={:?}", brillig.outputs);
+                for (pc, brillig_opcode) in brillig.bytecode.iter().enumerate() {
+                    println!(
+                        "{:>3}.{:<2} |{:2} {:?}",
+                        ip,
+                        pc,
+                        if pc == current_brillig_pc {
+                            marker
+                        } else if self.context.is_breakpoint_set(&OpcodeLocation::Brillig {
+                            acir_index: ip,
+                            brillig_index: pc,
+                        }) {
+                            " *"
+                        } else {
+                            ""
+                        },
+                        brillig_opcode
+                    );
```

Comment:
This print is doing a lot. Can you break it up to make it clear what these things are doing?

Reply:
Agreed! I also think that we need a better abstraction to deal with printing. I have a similar feeling when looking at the code that prints the current source location. I'll fiddle a bit and try to come up with something better.

Reply:
I iterated a bit and the result is much cleaner I think. I still feel there are some missing abstractions there, but I'm sure they will become evident with time.

Reply:
This looks much nicer now. Agreed on there being some more opportunities for abstraction but no need to rush into them.